### ISSUE TITLE
Lack of clone implementations makes some multithreading patterns difficult.

### ISSUE THREAD

**Issue by GenerallyIntelligent** [EventStore/EventStoreDB-Client-Rust#111]
TL;DR: Some structs, but most specifically `SubscriptionWrite` should derive `Clone` to allow for easy task per event patterns. An example use case with cloning is show here in this [gist](https://gist.github.com/GenerallyIntelligent/69aebd2570f7927d48a08d9e1442d7f7)

While writing some code using the client which needed to be multithreaded, I encountered difficulties with the `SubscriptionWrite` struct. I was delegating incoming events using `tokio::spawn`, and then handling them with some internal logic, after which I needed to ack them, but didn't want to send those events back somehow, or have the main thread do the acking at all.
Since it looks like this struct is a wrapper for an `mpsc::Sender`, I thought that I could use the common pattern of giving the task a clone the sender, but `Clone` is not implemented for `SubscriptionWrite`

**Comment by GenerallyIntelligent** [comment]
I did fork the repo and test out that adding `#[derive(Clone)]` didn't break anything, and it worked without any problem. Let me know if I should put together a pull request and I am more than happy to do so. Cheers!

**Comment by YoEight** [comment]
Hey @GenerallyIntelligent, thanks for creating your issue!

For some reason, I didn't see your post earlier, sorry. Subscriptions/PersistentSubscriptions changed quite a bit in the meantime. While those still do not implement the `Clone` trait, for persistent subscriptions, both the read and write parts have been merged. You can have a look at the related PR, there are also snippets of how those are used now: https://github.com/EventStore/EventStoreDB-Client-Rust/pull/110

While I am not against adding `Clone` per se (with the new design, it's no longer possible to implement clone anyway), your usage is not really how we usually recommend using persistent subscriptions. I would recommend connecting to a persistent subscription as many times as you need and using your subscription handle in the thread you want instead.

I remain open-minded on this topic and would like to know more about why you are using persistent subscriptions that way. Maybe I'm missing something in your use case.

**Comment by GenerallyIntelligent** [comment]
Thanks for getting back to me.

It does look like the new `PersistentSubscription` struct would make implementing clone practically impossible, since it contains the tonic stream.

As for why one would want to use persistent subscriptions this way:
We have some logic which we would like to be concurrent, because we don't want to impact our responsiveness to new events on the subscription. Because of this, we are spawning a new task to handle each event, and allowing the original task to block for the next event. (A common pattern with tokio, since the tasks are so cheap).
The issue we have is that our subscription is owned by the main task context (which is blocking for the next event), and we would like to ack the event we just handled. But since we are creating some arbitrary number of tasks, each of those tasks needs a way to ack its respective event.
There are a couple of ways to do this, but one solution is to give each task a clone of the subscription ack sender. This way, we are not creating new subscriptions (the main task is still blocking for the next event), but we are able to ack events within many different tasks concurrently. Both `tokio::sync::mpsc::Sender` and `tokio::sync::mpsc::UnboundedSender` implement clone, since it isn't important where we are sending from, and often helpful to delegate between tasks.

Please correct me if I am not understanding, but it sounds like you are suggesting we have some number of tasks for handling events, and connect to the persistent subscription on each, such that each task has a unique connection to the eventstore. The reason I am hesitant to do this is that it generates more overhead on the eventstore itself, and prevents us from able to scale as effectively to the volume of events.
If we receive a number of events equal to the number of tasks which we have created, we are unable to begin processing a new event, until all of the current events are processed, or until we can create a new connection to the persistent subscription. This can add significant latency in IO bound scenarios.
We are using a microservice architecture, and so will be scaling horizontally. I'd like to avoid creating 1000 subscriptions in separate tasks just so that we can handle a large number of events, and then scale that by 100 or so instances. I'm sure that the eventstore is robust, but we have a large number of services for which this sort of behavior would apply.

**Comment by YoEight** [comment]
> It does look like the new PersistentSubscription struct would make implementing clone practically impossible, since it contains the tonic stream.

The new design got rid of a lot of unneeded boxing, resulting in faster code overall.

> Please correct me if I am not understanding, but it sounds like you are suggesting we have some number of tasks for handling events, and connect to the persistent subscription on each, such that each task has a unique connection to the eventstore. 

You are right until the point where you thought each task needs its own connection to eventstore. An eventstore connection is cloneable. So you can create as many persistent subscriptions as you want with the same connection, and that's what I'm suggesting here. Obviously, you would have to use the right consumer strategy but it is worth investigating.

Overall, persistent subscriptions aren't as fast as regular subscriptions. Persistent subscriptions are only run by the leader node of a cluster for example. There is also the possibility of you not using the right feature too. Are you sure you need a persistent subscription anyway?

In the end, I can still sneak a way for you to have a detached handle to ack a persistent subscription event. However, I want to be sure the need for it is really founded.
